README.md


# BWStest

[![Build Status](https://github.com/shabbychef/BWStest/workflows/R-CMD-check/badge.svg)](https://github.com/shabbychef/BWStest/actions)
[![codecov.io](http://codecov.io/github/shabbychef/BWStest/coverage.svg?branch=master)](https://app.codecov.io/github/shabbychef/BWStest?branch=master)
[![CRAN](https://www.r-pkg.org/badges/version/BWStest)](https://cran.r-project.org/package=BWStest)
[![Downloads](http://cranlogs.r-pkg.org/badges/BWStest?color=green)](https://www.r-pkg.org:443/pkg/BWStest)
[![Total](http://cranlogs.r-pkg.org/badges/grand-total/BWStest?color=green)](https://www.r-pkg.org:443/pkg/BWStest)
![RCpp](https://img.shields.io/badge/RCpp-inside-blue.svg)

Performs the [Baumgartner-Weiß-Schindler 2-sample test](https://doi.org/10.2307/2533862) of equal probability
distributions. 

-- Steven E. Pav, shabbychef@gmail.com

## Installation

This package can be installed 
from CRAN, 
via [drat](https://github.com/eddelbuettel/drat "drat"), or
from github:


```r
# via CRAN:
install.packages("BWStest")
# via drat:
if (require(drat)) {
    drat:::add("shabbychef")
    install.packages("BWStest")
}
# get snapshot from github (may be buggy)
if (require(devtools)) {
    install_github("shabbychef/BWStest")
}
```

## What is it?

The Baumgartner-Weiß-Schindler ('BWS') test is a non-parametric hypothesis test for the null of
equal probability distributions of two samples, not unlike the two-sample
Kolmogorov-Smirnov test or the Wilcoxon test. 
The BWS test is more powerful than these other tests under certain
alternatives, as shown in the original paper and replicated below.

## Basic Usage

The front end for the hypothesis test is the function `bws_test`. By default it supports
the the classical Baumgartner-Weiß-Schindler test for a two-sided alternative, returning a `htest` object:


```r
require(BWStest)
set.seed(12345)
# under the null:
x <- rnorm(200)
y <- rnorm(200)
hval <- bws_test(x, y)
show(hval)
```

```
## 
## 	two-sample BWS test
## 
## data:  x vs. y
## B = 1, p-value = 0.2
## alternative hypothesis: true difference in survival functions is not equal to 0
```

```r
# under the alternative:
z <- rnorm(200, mean = 1)
hval <- bws_test(x, z)
show(hval)
```

```
## 
## 	two-sample BWS test
## 
## data:  x vs. z
## B = 30, p-value <2e-16
## alternative hypothesis: true difference in survival functions is not equal to 0
```

The code also supports alternative test statistics from Neuhäuser and Murakami, along with supporting
one-sided alternatives for some of the tests:


```r
set.seed(12345)
# under the alternative:
x <- rnorm(200)
y <- rnorm(200, mean = 1)
hval <- bws_test(x, z, alternative = "less")
show(hval)
```

```
## 
## 	two-sample Neuhauser/Murakami test
## 
## data:  x vs. z
## B_2 = -30, p-value <2e-16
## alternative hypothesis: true difference in survival functions is less than 0
```

```r
x <- rnorm(8)
y <- rnorm(8, mean = 1)
hval <- bws_test(x, z, method = "B3", alternative = "two.sided")
show(hval)
```

```
## 
## 	two-sample Murakami test
## 
## data:  x vs. z
## B_3 = 2, p-value = 0.009
## alternative hypothesis: true difference in survival functions is not equal to 0
```

We should note that the `B3` through `B5` tests do _not_ achieve nominal coverage
for large sample sizes and should _only_ be used on sample sizes of about 12 or fewer
in each of the two samples.

## Checking the null

  _Doverai No Proverai_ (Trust, but verify.) -- Russian proverb.

Here we perform 5000 simulations of the BWS test under the null hypothesis, then
compute the CDF of the test statistic. If the code is correct, the resultant p-values
should be uniform. So I q-q plot under the uniform law:


```r
require(BWStest)

# now compute a bunch under the null:
set.seed(1234)
bvals <- replicate(5000, bws_stat(rnorm(100), rnorm(100)))
# compute the approximate p-values under the null:
pvals <- bws_cdf(bvals)

require(ggplot2)
ph <- ggplot(data.frame(pv = pvals), aes(sample = pv)) + 
    stat_qq(distribution = stats::qunif)
print(ph)
```

<img src="tools/figure/babysteps-1.png" title="plot of chunk babysteps" alt="plot of chunk babysteps" width="500px" height="400px" />

Looks good to me!

## Under the alternative

Here we replicate figure 2A of Baumgartner _et al._ We draw two samples from the normal distribution,
both with unit standard deviation, letting _a_ be the difference in means. 
We check the empirical rejection rate at the 0.05 level for a few different tests.
As in Baumgartner, we find that the lowly t-test
is the most powerful in this case, with the BWS, Cramer-Von Mises, and Wilcoxon tests displaying similar power, 
then the KS test the least powerful. Note that the Kolmogorov Smirnov test does not appear to have nominal
coverage under the null, probably due to the small sample size.


```r
n.sim <- 10000
avals <- seq(0, 3.2, length.out = 17)
alpha <- 0.05
mnsize <- 10

# this is archived on CRAN, unfortunately:
library(CvM2SL2Test)
# find the CVM critical value.
critv <- uniroot(function(x) {
    cvmts.pval(x, mnsize, mnsize) - alpha
}, lower = 0, upper = 100, maxiter = 5000)$root

set.seed(1234)
simul <- sapply(avals, function(a) {
    rejs <- replicate(n.sim, {
        x <- rnorm(mnsize, mean = 0)
        y <- rnorm(mnsize, mean = a)
        bws <- bws_cdf(bws_stat(x, y), lower_tail = FALSE) <= 
            alpha
        ttv <- t.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        cvm <- cvmts.test(x, y) >= critv
        ksv <- ks.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        wcx <- wilcox.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        c(bws, ttv, cvm, ksv, wcx)
    })
    rejrate <- rowMeans(rejs)
    names(rejrate) <- c("BWS test", "t test", "Cramer-Von Mises test", 
        "Kolmogorov Smirnov test", "Wilcoxon test")
    rejrate
}, simplify = "matrix")

Arejrates <- data.frame(t(simul))
Arejrates$a <- avals
```

```r
library(tidyr)
library(dplyr)
library(ggplot2)
plotdf <- tidyr::gather(Arejrates, "test", "rejection_rate", 
    -a) %>% dplyr::mutate(test = gsub("\\.", " ", test))
ph <- ggplot(plotdf, aes(x = a, y = rejection_rate, 
    group = test, colour = test)) + geom_line() + geom_point() + 
    labs(x = "a, difference in means", y = "rejection rate")
print(ph)
```

<img src="tools/figure/fig_two_A-1.png" title="plot of chunk fig_two_A" alt="plot of chunk fig_two_A" width="700px" height="500px" />

Here we replicate figure 2B of Baumgartner _et al._ We draw two samples from the normal distribution,
both with zero mean, one with unit standard deviation, the other with standard deviation of _sigma_.
We compute the empirical rejection rate at the 0.05 level, dropping the t-test
since it is not relevant for this formulation.
As in Baumgartner, 
we find the BWS test is the most powerful, followed by KS test, then Cramer-Von Mises, 
then Wilcoxon, which is basically useless in this simulation.


```r
n.sim <- 10000
svals <- seq(1, 45, length.out = 10)
alpha <- 0.05
mnsize <- 10

# this is archived on CRAN, unfortunately:
library(CvM2SL2Test)
# find the CVM critical value.
critv <- uniroot(function(x) {
    cvmts.pval(x, mnsize, mnsize) - alpha
}, lower = 0, upper = 100, maxiter = 5000)$root

set.seed(1234)
simul <- sapply(svals, function(s) {
    rejs <- replicate(n.sim, {
        x <- rnorm(mnsize, mean = 0, sd = 1)
        y <- rnorm(mnsize, mean = 0, sd = s)
        bws <- bws_cdf(bws_stat(x, y), lower_tail = FALSE) <= 
            alpha
        cvm <- cvmts.test(x, y) >= critv
        ksv <- ks.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        wcx <- wilcox.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        c(bws, cvm, ksv, wcx)
    })
    rejrate <- rowMeans(rejs)
    names(rejrate) <- c("BWS test", "Cramer-Von Mises test", 
        "Kolmogorov Smirnov test", "Wilcoxon test")
    rejrate
}, simplify = "matrix")

Brejrates <- data.frame(t(simul))
Brejrates$sigma <- svals
```

```r
plotdf <- tidyr::gather(Brejrates, "test", "rejection_rate", 
    -sigma) %>% dplyr::mutate(test = gsub("\\.", " ", 
    test))
ph <- ggplot(plotdf, aes(x = sigma, y = rejection_rate, 
    group = test, colour = test)) + geom_line() + geom_point() + 
    labs(x = "sigma, ratio of standard deviations", 
        y = "rejection rate")
print(ph)
```

<img src="tools/figure/fig_two_B-1.png" title="plot of chunk fig_two_B" alt="plot of chunk fig_two_B" width="700px" height="500px" />

Here we replicate figure 3A of Baumgartner _et al._ We draw two samples from the exponential distribution,
letting _l_ be the ratio of the rate parameters of the two populations.
We compute the empirical rejection rate at the 0.05 level.
As in Baumgartner, 
we find the BWS test is the most powerful, followed by Wilcoxon, then Cramer-Von Mises, then the
KS test.


```r
n.sim <- 10000
lvals <- seq(1, 12)
alpha <- 0.05
mnsize <- 10

# this is archived on CRAN, unfortunately:
library(CvM2SL2Test)
# find the CVM critical value.
critv <- uniroot(function(x) {
    cvmts.pval(x, mnsize, mnsize) - alpha
}, lower = 0, upper = 100, maxiter = 5000)$root

set.seed(1234)
simul <- sapply(lvals, function(l) {
    rejs <- replicate(n.sim, {
        x <- rexp(mnsize, rate = 1)
        y <- rexp(mnsize, rate = l)
        bws <- bws_cdf(bws_stat(x, y), lower_tail = FALSE) <= 
            alpha
        cvm <- cvmts.test(x, y) >= critv
        ksv <- ks.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        wcx <- wilcox.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        c(bws, cvm, ksv, wcx)
    })
    rejrate <- rowMeans(rejs)
    names(rejrate) <- c("BWS test", "Cramer-Von Mises test", 
        "Kolmogorov Smirnov test", "Wilcoxon test")
    rejrate
}, simplify = "matrix")

Crejrates <- data.frame(t(simul))
Crejrates$lratio <- lvals
```

```r
plotdf <- tidyr::gather(Crejrates, "test", "rejection_rate", 
    -lratio) %>% dplyr::mutate(test = gsub("\\.", " ", 
    test))
ph <- ggplot(plotdf, aes(x = lratio, y = rejection_rate, 
    group = test, colour = test)) + geom_line() + geom_point() + 
    labs(x = "l, ratio of rate parameters", y = "rejection rate")
print(ph)
```

<img src="tools/figure/fig_three_A-1.png" title="plot of chunk fig_three_A" alt="plot of chunk fig_three_A" width="700px" height="500px" />

Here we replicate figure 3B of Baumgartner _et al._ We draw two samples, one from the normal distribution
with zero mean and variance one-twelth, the other uniformly on -0.5 to 0.5. We take equal sample sizes
from these two populations, then vary the sample size, checking the
empirical rejection rate at the 0.05 level. Since the first two moments are equal, the Wilcoxon test
is useless here, and not applied.
As in Baumgartner, we find the BWS test is the most powerful, 
followed by the KS test and Cramer-Von Mises tests. 
Based on the power plots here, I theorize that Baumgartner _et al._ are plotting the _total_ sample
sizes on the _x_ axis, that is, drawing _n_ from both distributions, then plotting empirical power
versus _2n_. We follow that convention, which makes the plots match those of Baumgartner.


```r
n.sim <- 10000
mvals <- seq(10, 670, by = 60)
alpha <- 0.05

# this is archived on CRAN, unfortunately:
library(CvM2SL2Test)

set.seed(1234)
simul <- sapply(mvals, function(mnsize) {
    # find the CVM critical value.  note that this
    # basically converged for mnsize > 100 or so, so we
    # take the min..  for reasons of speed.
    critv <- uniroot(function(x) {
        cvmts.pval(x, min(mnsize, 80), min(mnsize, 
            80)) - alpha
    }, lower = 0, upper = 2, maxiter = 100)$root
    rejs <- replicate(n.sim, {
        x <- rnorm(mnsize, mean = 0, sd = 1/sqrt(12))
        y <- runif(mnsize, min = -0.5, max = 0.5)
        bws <- bws_cdf(bws_stat(x, y), lower_tail = FALSE) <= 
            alpha
        cvm <- cvmts.test(x, y) >= critv
        ksv <- ks.test(x, y, alternative = "two.sided")$p.value <= 
            alpha
        c(bws, cvm, ksv)
    })
    rejrate <- rowMeans(rejs)
    names(rejrate) <- c("BWS test", "Cramer-Von Mises test", 
        "Kolmogorov Smirnov test")
    rejrate
}, simplify = "matrix")

Drejrates <- data.frame(t(simul))
Drejrates$ssize <- 2 * mvals
```

```r
plotdf <- tidyr::gather(Drejrates, "test", "rejection_rate", 
    -ssize) %>% dplyr::mutate(test = gsub("\\.", " ", 
    test))
ph <- ggplot(plotdf, aes(x = ssize, y = rejection_rate, 
    group = test, colour = test)) + geom_line() + geom_point() + 
    labs(x = "m+n, total sample size", y = "rejection rate")
print(ph)
```

<img src="tools/figure/fig_three_B-1.png" title="plot of chunk fig_three_B" alt="plot of chunk fig_three_B" width="700px" height="500px" />

## Murakami tests

[Neuhäuser](https://doi.org/10.1007/BF02762032) and 
[Murakami](https://doi.org/10.1080/00949655.2010.551516) described
some modifications to the original test of Baumgartner. Neuhäuser's
test allows one to test against directional alternatives. Murakami
enumerated some modifications to the weighting scheme. These are
available via the `murakami_stat` function, where the `flavor`
corresponds to the test number from Murakami's paper, namely

* 0 corresponds to the original BWS statistic.
* 1 corresponds to Murakami's first modification.
* 2 corresponds to Neuhäuser's statistic, which can take negative values.
* 3 through 5 correspond to different weighting schemes of Murakami's.

Here we take these through the paces as above.

### Under the null

As above, we draw samples under the null and compare to the putative
CDF function


```r
require(BWStest)

# now compute a bunch under the null:
n1 <- 9
n2 <- n1
set.seed(1234)
allpvs <- lapply(0L:5L, function(flavor) {
    bvals <- replicate(5000, murakami_stat(rnorm(n1), 
        rnorm(n2), flavor = flavor))
    # compute the approximate p-values under the null:
    pvals <- murakami_cdf(bvals, n1 = n1, n2 = n2, 
        flavor = flavor)
    data.frame(pv = pvals, flavor = flavor)
})

df <- do.call(rbind, allpvs)

require(ggplot2)
ph <- ggplot(df, aes(sample = pv)) + facet_grid(flavor ~ 
    .) + stat_qq(distribution = stats::qunif)
print(ph)
```

<img src="tools/figure/murakami_null-1.png" title="plot of chunk murakami_null" alt="plot of chunk murakami_null" width="500px" height="600px" />

While these all look fine, they are based on small sample sizes. The CDF is approximated by evaluating
all the permutations (with memoisation to tame the computational requirements), but this can only be
done up to some reasonably small sample size. If the test statistic does not converge beyond that
sample size, the CDF approximation will not be accurate. This appears to be the case for flavors 3 through 5,
as demonstrated below:


```r
require(BWStest)

# now compute a bunch under the null:
n1 <- 50
n2 <- n1
set.seed(1234)
allpvs <- lapply(0L:5L, function(flavor) {
    bvals <- replicate(5000, murakami_stat(rnorm(n1), 
        rnorm(n2), flavor = flavor))
    # compute the approximate p-values under the null:
    pvals <- murakami_cdf(bvals, n1 = n1, n2 = n2, 
        flavor = flavor)
    data.frame(pv = pvals, flavor = flavor)
})

df <- do.call(rbind, allpvs)

require(ggplot2)
ph <- ggplot(df, aes(sample = pv)) + facet_grid(flavor ~ 
    .) + stat_qq(distribution = stats::qunif)
print(ph)
```

<img src="tools/figure/murakami_null_2-1.png" title="plot of chunk murakami_null_2" alt="plot of chunk murakami_null_2" width="500px" height="600px" />

So that's not so great. 

## Under the alternative

We can perform the power comparisons of Baumgartner _et al._ using the Murakami test statistics.
Here we consider the setup of figure 2A, samples of size 10 from the normal distribution with
equal variances but different means. Neuhäuser's test is, as expected, more powerful for detecting
this one-sided alternative, probably on par with the t-test, while the BWS test (basically equivalent
to Murakami 1) is next, then 5, 3, 4.


```r
n.sim <- 10000
avals <- seq(0, 3.2, length.out = 17)
alpha <- 0.05
mnsize <- 10

set.seed(1234)
simul <- sapply(avals, function(a) {
    statvs <- replicate(n.sim, {
        x <- rnorm(mnsize, mean = 0)
        y <- rnorm(mnsize, mean = a)
        bws <- bws_stat(x, y)
        murav <- sapply(1L:5L, function(flavor) {
            murakami_stat(x, y, flavor = flavor)
        })
        c(bws, murav)
    })
    rej_bws <- mean(bws_cdf(statvs[1, ], lower_tail = FALSE) <= 
        alpha)
    rej_mur <- sapply(2:6, function(rown) {
        mean(murakami_cdf(statvs[rown, ], mnsize, mnsize, 
            flavor = rown - 1, lower_tail = FALSE) <= 
            alpha)
    })
    
    rejrate <- c(rej_bws, rej_mur)
    names(rejrate) <- c("BWS test", "Murakami 1", "Neuhauser", 
        "Murakami 3", "Murakami 4", "Murakami 5")
    rejrate
}, simplify = "matrix")

Arejrates <- data.frame(t(simul))
Arejrates$a <- avals
```

```r
library(tidyr)
library(dplyr)
library(ggplot2)
plotdf <- tidyr::gather(Arejrates, "test", "rejection_rate", 
    -a) %>% dplyr::mutate(test = gsub("\\.", " ", test))
ph <- ggplot(plotdf, aes(x = a, y = rejection_rate, 
    group = test, colour = test)) + geom_line() + geom_point() + 
    labs(x = "a, difference in means", y = "rejection rate")
print(ph)
```

<img src="tools/figure/murakami_fig_two_A-1.png" title="plot of chunk murakami_fig_two_A" alt="plot of chunk murakami_fig_two_A" width="700px" height="500px" />

Now we consider figure2B, with two samples of equal size from the normal distribution with zero

ChangeLog

2023-10-10 Steven E. Pav <shabbychef@gmail.com>
* 0.2.3 on CRAN
* Update doi links. 

2018-10-17 Steven E. Pav <shabbychef@gmail.com>
* 0.2.2 on CRAN
* Package maintenance--no new features.

2017-03-20 Steven E. Pav <shabbychef@gmail.com>
* 0.2.1 on CRAN
* move github figures to location CRAN understands.
* package initialization mumbo jumbo, see Rcpp issue 636.

2016-04-29 Steven E. Pav <shabbychef@gmail.com>
* 0.2.0 on CRAN
* Adding Murakami statistics.

2016-04-07 Steven E. Pav <shabbychef@gmail.com>
* 0.1.0 on CRAN

2016-04-06 Steven E. Pav <shabbychef@gmail.com>
* start work



src/murakami_stat.cpp
/*
 
  This file is part of BWStest.
  
  BWStest is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  BWStest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.
  
  You should have received a copy of the GNU Lesser General Public License
  along with BWStest.  If not, see <http://www.gnu.org/licenses/>.

  Baumgartner-Weiss-Schindler 2-sample test of equal distributions.
 
  see also: 

  + W. Baumgartner, P. Weiss, H. Schindler, 'A Nonparametric Test for the General Two-Sample Problem', 
    Biometrics, Vol. 54, No. 3 (Sep., 1998), pp. 1129-1135. \doi{10.2307/2533862}
  + M. Neuhauser, 'Exact Tests Based on the Baumgartner-Weiss-Schindler Statistic--A Survey', 
    Statistical Papers, Vol 46 (2005), pp. 1-30. \doi{10.1007/BF02762032}
  + M. Neuhauser, 'One-Sided Two-Sample and Trend Tests Based on a Modified Baumgartner-Weiss-Schindler 
    Statistic', J. Nonparametric Statistics, Vol 13 (2001) pp 729-739. \doi{10.1080/10485250108832874}
  + H. Murakami, 'K-Sample Rank Test Based on Modified Baumgartner Statistic and its Power Comparison', 
    J. Jpn. Comp. Statist. Vol 19 (2006), pp. 1-13. \doi{10.1080/00949655.2010.551516}


  Created: 2016.04.06
  Copyright: Steven E. Pav, 2016
  Author: Steven E. Pav <steven@corecast.io>
  Comments: Steven E. Pav
*/

#ifndef __DEF_MURAKAMI_STAT__
#define __DEF_MURAKAMI_STAT__

#include <math.h>
#include "util.h"

// 2FIX: deal with NA/NAN here?
#define MAX(a,b) ((a>b)? (a):(b))
#define MIN(a,b) ((a<b)? (a):(b))

#endif /* __DEF_MURAKAMI_STAT__ */

#include <Rcpp.h>
using namespace Rcpp;

// a weird interface, but basically take the sample_set
// and compute a bunch of the Murakami B_fx statistics,
// for f=0,1,2,3,4,5, set by the 'flavor'. We increment
// through a bunch of sample_sets for efficiency.
//
// later make this back to a template to squeeze out the last bits of efficiency?
//template <int flavor>
//NumericVector murakami_pre_B(const size_t N,const size_t nx,IntegerVector parts,const size_t numits) {
NumericVector murakami_pre_B(const size_t N,const size_t nx,IntegerVector parts,const size_t numits,const int flavor) {
	if ((flavor < 0) || (flavor > 5)) { stop("unsupported flavor."); }
	int iii,jjj,nnn;
	double evx,vvx,Np1,nxp1;
	double nonce,npart,dpart,bplus;
	// preallocate
	NumericVector B1(numits);
	const int ny = (int)N - (int)nx;

	Np1  = (double)N + 1.0;
	nxp1 = (double)nx + 1.0;

	switch(flavor) {
		case 0:
		case 2:
			evx = ((double)N) / ((double)nx);
			vvx = ((double)ny) * evx;
			break;
		default:
			evx = Np1 / nxp1;
			vvx = ((double)ny * Np1) / ((double)nx + 2.0);
			break;
	}

	//if ((flavor == 0) || (flavor == 2)) {
		//evx = (double)N / (double)nx;
		//vvx = (double)ny * evx;
	//}
	//if ((flavor == 1) || (flavor == 3) || (flavor == 4) || (flavor == 5)) {
		//evx = Np1 / nxp1;
		//vvx = (double)ny * Np1 / ((double)nx + 2.0);
	//}
	for (jjj=0;jjj<numits;jjj++) {
		// for more on partitions nonsense, see also:
		// http://howardhinnant.github.io/combinations.html
		// note this will not scale to the multi-class problem very well...
		if (jjj > 0) {
			// increment
			iii = nx - 1;
			while ((iii > 0) && (parts(iii) == N - (nx - 1) + iii)) {
				iii--;
			}
			parts(iii)++;
			for (;iii < (nx - 1);iii++) {
				parts(iii+1) = parts(iii)+1;
			}
		}
		// just to be sure we have prealloc'ed with zeros:
		B1(jjj) = 0.0;
		for (iii=1;iii<=nx;iii++) {
			nnn = parts(iii-1);
			npart = ((double)nnn - evx * (double)iii);
			nonce = ((double)iii) / nxp1;
			dpart = nonce * (1.0 - nonce) * vvx;

			switch(flavor) {
				case 0:
				case 1:
					bplus = npart * npart / dpart;
					break;
				case 2:
					bplus = (npart * std::abs(npart)) / dpart;
					break;
				case 3:
					bplus = npart * npart / (dpart * dpart);
					break;
				case 4:
					bplus = std::abs(npart) / (dpart * dpart);
					break;
				case 5:
					bplus = npart * npart / log(dpart);
					break;
			}
			B1(jjj) += bplus / ((double) nx);
		}
	}
	return B1;
}

//// from when it was templated...
//// same interface, not templated, supposedly faster because of it
//NumericVector murakami_B(const size_t N,const size_t nx,IntegerVector parts,const size_t numits,int flavor) {
	//NumericVector B1;
	//switch(flavor) {
		//case 0:
			//B1 = murakami_pre_B<0>(N,nx,parts,numits);
			//break;
		//case 1:
			//B1 = murakami_pre_B<1>(N,nx,parts,numits);
			//break;
		//case 2:
			//B1 = murakami_pre_B<2>(N,nx,parts,numits);
			//break;
		//case 3:
			//B1 = murakami_pre_B<3>(N,nx,parts,numits);
			//break;
		//case 4:
			//B1 = murakami_pre_B<4>(N,nx,parts,numits);
			//break;
		//case 5:
			//B1 = murakami_pre_B<5>(N,nx,parts,numits);
			//break;
	//}
	//return B1;
//}


/*
double murakami_raw(int N,NumericVector x,int flavor=0) {
	const size_t nx = (size_t)x.size();
	IntegerVector parts(nx);
	for (int iii=0;iii<nx;iii++) { parts(iii) = (int)x(iii); }
	NumericVector B1 = murakami_pre_B(N,nx,parts,1,flavor);
	return B1(0);
}
*/

NumericVector murakami_many_B(const int N,const int nx,int flavor) {
	IntegerVector parts(nx);
	int iii;
	for (iii=0;iii<nx;iii++) {
		// our partitions are 1-based.
		parts(iii)=iii+1;
	}
	size_t numits = (size_t)Combinatorics::bincoef[N][nx];
	return murakami_pre_B(N,nx,parts,numits,flavor);
}

//' @title
//' Compute Murakami's test statistic.
//'
//' @description
//'
//' Compute one of the modified Baumgartner-Weiss-Schindler test statistics proposed
//' by Murakami, or Neuhauser.
//'
//' @details
//'
//' Given vectors \eqn{X} and \eqn{Y}, computes \eqn{B_{jX}} and \eqn{B_{jY}} 
//' for some \eqn{j} as described by Murakami and by Neuhauser, returning either their 
//' their average or their average distance.
//' The test statistics approximate the weighted square norm of the
//' difference in CDFs of the two distributions. 
//'
//' The test statistic is based only on the ranks of the input. If the same
//' monotonic transform is applied to both vectors, the result should be unchanged.
//'
//' The various \sQuote{flavor}s of test statistic are:
//' \describe{
//' \item{0}{The statistic of Baumgartner-Weiss-Schindler.}
//' \item{1}{Murakami's \eqn{B_1} statistic, from his 2006 paper.}
//' \item{2}{Neuhauser's difference statistic, denoted by Murakami as \eqn{B_2} in his 
//' 2012 paper.}
//' \item{3}{Murakami's \eqn{B_3} statistic, from his 2012 paper.}
//' \item{4}{Murakami's \eqn{B_4} statistic, from his 2012 paper.}
//' \item{5}{Murakami's \eqn{B_5} statistic, from his 2012 paper, with a log weighting.}
//' }
//'
//' @param x a vector of the first sample.
//' @param y a vector of the second sample.
//' @param flavor which \sQuote{flavor} of test statistic. 
//' @param nx the length of \code{x}, the first sample.
//' @param ny the length of \code{y}, the second sample.
//'
//' @return The BWS test statistic, \eqn{B_j}. For \code{murakami_stat_perms}, a vector of
//' the test statistics for \emph{all} permutations of the input.
//' @note \code{NA} and \code{NaN} are not yet dealt with!
//' @seealso \code{\link{bws_stat}}.
//' @examples
//'
//' set.seed(1234)
//' x <- runif(1000)
//' y <- runif(100)
//' bval <- murakami_stat(x,y,1)
//'
//' \donttest{
//' nx <- 6
//' ny <- 5
//' # monte carlo
//' set.seed(1234)
//' repli <- replicate(3000,murakami_stat(rnorm(nx),rnorm(ny),0L))
//' # under the null, perform the permutation test:
//' allem <- murakami_stat_perms(nx,ny,0L)
//' plot(ecdf(allem)) 
//' lines(ecdf(repli),col='red') 
//' }
//'
//' @template etc
//' @template ref-bws
//' @template ref-modtests
//' @rdname murakami_stat
//' @export
// [[Rcpp::export]]
double murakami_stat(NumericVector x,NumericVector y,int flavor=0) {
	// 2FIX: add na_rm?
	// put into the form to be consumed by the other function...
	NumericVector sortx = clone(x); std::sort(sortx.begin(), sortx.end());
	NumericVector sorty = clone(y); std::sort(sorty.begin(), sorty.end());
	IntegerVector G = full_rank<NumericVector, double>(sorty, sortx);
	IntegerVector H = full_rank<NumericVector, double>(sortx, sorty);

	const size_t nx = (size_t)x.size();
	const size_t ny = (size_t)y.size();
	const size_t N = nx + ny;

	NumericVector B1 = murakami_pre_B(N,nx,G,1,flavor);
	NumericVector B2 = murakami_pre_B(N,ny,H,1,flavor);
	double B;

	switch(flavor) {
		case 2:
			B = 0.5 * (B2(0) - B1(0));
			break;
		default:
			B = 0.5 * (B2(0) + B1(0));
			break;
	}
	return B;
}
//' @rdname murakami_stat
//' @export
// [[Rcpp::export]]
NumericVector murakami_stat_perms(int nx, int ny,int flavor=0) {
	const int N = nx + ny;
	if (N > Combinatorics::BINCOEF_MAX_ORD) { stop("N too large"); }

	const bool IS_SYMMETRIC = (nx == ny);
	NumericVector B1,B2,B;
	B1 = murakami_many_B(N,nx,flavor);
	int nlen=B1.size();
	B = NumericVector(nlen);
	int iii;
	if (IS_SYMMETRIC) {
		// combine them;
		if (flavor == 2) {
			for (iii=0;iii < nlen;iii++) {
				B(iii) = 0.5 * (B1(nlen-iii-1) - B1(iii));
			}
		} else {
			for (iii=0;iii < nlen;iii++) {
				B(iii) = 0.5 * (B1(nlen-iii-1) + B1(iii));
			}
		}
	} else {
		B2 = murakami_many_B(N,ny,flavor);
		// combine them;
		if (flavor == 2) {
			for (iii=0;iii < nlen;iii++) {
				B(iii) = 0.5 * (B2(nlen-iii-1) - B1(iii));
			}
		} else {
			for (iii=0;iii < nlen;iii++) {
				B(iii) = 0.5 * (B2(nlen-iii-1) + B1(iii));
			}
		}
	}
	return B;
}

// for speed comparison, timings of the new method defined here:
//    a <- murakami_all_B(5, 5, 1)     24.12     25.35     29.67     26.77     32.91     51.08   100
//    b <- murakami_all_B(7, 7, 1)    366.92    370.64    420.82    379.81    456.58   1235.38   100
//   cc <- murakami_all_B(9, 9, 1)   7106.97   7192.79   7588.08   7304.08   7566.63  10396.69   100
// ee <- murakami_all_B(11, 11, 1) 125301.28 126489.78 135681.89 127329.02 149259.48 183445.64   100
// and the old method based on partitions::setparts and whatnot:
// aa <- murakami_BB(5, 5, 1)     646.9     719.4     835.2     818.4     885.1    1636   100
// bb <- murakami_BB(7, 7, 1)   20391.0   20632.6   21556.7   20869.3   21993.5   26766   100
// cc <- murakami_BB(9, 9, 1) 1374527.5 1405699.7 1474427.9 1449461.6 1497241.3 2010725   100
// I never got up to 11 & 11 with the old method.
// the takeaway is that we are much faster now. yay.
// 
//

/*

		 3/2 split:
 
		 123 45
		 124 35
		 125 34
		 134 25
		 135 24
		 145 23
		 234 15
		 235 14
		 245 13
		 345 12


		 2/3 split:

		 12 345
		 13 245
		 14 235
		 15 234
		 23 145
		 24 135
		 25 134
		 34 124
		 35 124
		 45 123

		 require(partitions)
		 M <- partitions::setparts(c(9,6)) 
		 allem <- sapply(seq_len(ncol(M)),function(ccc) {
		 		xv <- which(M[,ccc] == 1)
		 		yv <- which(M[,ccc] == 2)
				murakami_stat(xv,yv,flavor=2)
		 })

		 require(partitions)
		 M <- partitions::setparts(c(3,2)) 
		 allem <- sapply(seq_len(ncol(M)),function(ccc) {
		 		xv <- which(M[,ccc] == 1)
		 		yv <- which(M[,ccc] == 2)
				murakami_stat(xv,yv,flavor=2)
		 })

		 sort(allem)

		 fooz <- c(murakami_stat(c(1,2,3),c(4,5),flavor=2),
		 murakami_stat(c(1,2,4),c(3,5),flavor=2),
		 murakami_stat(c(1,2,5),c(3,4),flavor=2),
		 murakami_stat(c(1,3,4),c(2,5),flavor=2),
		 murakami_stat(c(1,3,5),c(2,4),flavor=2),
		 murakami_stat(c(1,4,5),c(2,3),flavor=2),
		 murakami_stat(c(2,3,4),c(1,5),flavor=2),
		 murakami_stat(c(2,3,5),c(1,4),flavor=2),
		 murakami_stat(c(2,4,5),c(1,3),flavor=2),
		 murakami_stat(c(3,4,5),c(1,2),flavor=2))

	sort(fooz) - sort(allem)
	sort(murakami_stat_perms(3,2,flavor=2)) 



 */


//for vim modeline: (do not edit)
// vim:nowrap:ts=2:sw=2:tw=129:fdm=marker:fmr=FOLDUP,UNFOLD:cms=//%s:tags=.c_tags;:syn=cpp:ft=cpp:mps+=<\:>:ai:si:cin:nu:fo=croql:cino=p0t0c5(0:
